File: Server/assets/reserve-workflow.js

// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.

;(function () {

var ReserveWorkflowSelection = Backbone.Model.extend({
});

window.ReserveWorkflow = Backbone.View.extend({
    template: JST['reserve-workflow'],
    events: {
        'submit form': 'submit',
        'change .system-picker input, select': 'update_selection',
        'change .job-options input': 'update_selection',
    },
    initialize: function (options) {
        this.request_in_progress = false;
        this.distro_picker = new DistroPicker({
            osmajor: options.options.osmajor,
            tag: options.options.tag,
            distro: options.options.distro,
            distro_tree_id: options.options.distro_tree_id,
            selection: options.selection,
        });
        this.labs = options.options['lab'];
        var default_pick = (options.selection['system'] ? 'fqdn' : 'auto');
        this.selection = new ReserveWorkflowSelection({
            lab: options.selection['lab'] || this.labs[0],
            system: options.selection['system'],
            pick: options.selection['pick'] || default_pick,
            reserve_duration: options.selection['reserve_duration'] || (24 * 60 * 60),
        });
        this.reserve_input = new ReservationDurationSelection(
            {reserve_duration: this.selection.get('reserve_duration')});
        // sync our selection from distro picker's selection
        this.listenTo(this.distro_picker.selection, 'change', function (model) {
            this.selection.set('distro_tree_id', model.get('distro_tree_id'));
        });
        this.listenTo(this.distro_picker.selection, 'change', this.update_query_string);
        this.listenTo(this.selection, 'change', this.update_button_state);
        this.render();
    },
    render: function () {
        this.$el.html(this.template(this));
        this.$('.job-options > legend').after(this.reserve_input.$el);
        this.reserve_input.render();
        this.distro_picker.setElement(this.$('.distro-picker')).render();
        this.update_button_state();
    },
    update_query_string: function () {
        // Use window.replaceState to add query parameters for distro tree 
        // selection, so that if the user comes back to this page the 
        // distro_tree_id <select/> will be populated without needing to wait for 
        // an AJAX request.
        var qs = new URI(window.location.search);
        if (this.distro_picker.selection.get('osmajor'))
            qs.setSearch('osmajor', this.distro_picker.selection.get('osmajor'));
        if (this.distro_picker.selection.get('tag'))
            qs.setSearch('tag', this.distro_picker.selection.get('tag'));
        if (this.distro_picker.selection.get('distro'))
            qs.setSearch('distro', this.distro_picker.selection.get('distro'));
        window.history.replaceState(undefined, undefined, qs.toString());
    },
    update_button_state: function () {
        var distro_tree_selected = ((this.selection.get('distro_tree_id') || []).length > 0);
        // enable/disable submit button
        this.$('.form-actions button').prop('disabled',
                (this.request_in_progress || !distro_tree_selected));
        // update href for system picker
        var href = new URI('../reserve_system');
        if (distro_tree_selected)
            href.setSearch('distro_tree_id', this.selection.get('distro_tree_id')[0]);
        this.$('a.select-system').attr('href', href.toString());
    },
    update_selection: function (evt) {
        var elem = evt.currentTarget;
        this.selection.set(elem.name, $(elem).val());
    },
    submit: function (evt) {
        evt.preventDefault();
        this.selection.set('reserve_duration', this.reserve_input.get_reservation_duration().asSeconds());
        var xhr = $.ajax({
            url: 'doit',
            type: 'POST',
            data: this.selection.attributes,
            traditional: true,
        });
        this.request_in_progress = true;
        this.$('.submit-status').html('<i class="fa fa-spinner fa-spin"></i> Submitting&hellip;');
        this.update_button_state();
        this.$('.form-actions .alert-error').remove();
        xhr.done(function (result) {
            // don't re-enable the button in the success case, we want the user 
            // to wait for their browser to load the job page
            window.location = xhr.getResponseHeader('Location');
        });
        xhr.fail(_.bind(function () {
            this.request_in_progress = false;
            this.$('.submit-status').empty();
            this.update_button_state();
            this.$('.form-actions').prepend(alert_for_xhr(xhr, 'Failed to submit job'));
        }, this));
    },
});

})();

File: Server/bkr/server/rpc/reserve_workflow.py

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

import six
from turbogears import expose
from bkr.server import identity
from bkr.server.bexceptions import DatabaseLookupError
from bkr.server.model import (Distro, Arch, OSMajor, DistroTag, OSVersion,
        DistroTree, LabController, System)


class ReserveWorkflow:

    @identity.require(identity.not_anonymous())
    @expose(template='bkr.server.templates.reserve_workflow')
    def index(self, **kwargs):
        # CherryPy will give us distro_tree_id as a scalar if it only has one
        # value, but we want it to always be a list of int
        if not kwargs.get('distro_tree_id'):
            kwargs['distro_tree_id'] = []
        elif not isinstance(kwargs['distro_tree_id'], list):
            kwargs['distro_tree_id'] = [int(kwargs['distro_tree_id'])]
        else:
            kwargs['distro_tree_id'] = [int(x) for x in kwargs['distro_tree_id']]

        # If we got a distro_tree_id but no osmajor or distro, fill those in
        # with the right values so that the distro picker is populated properly
        if kwargs['distro_tree_id']:
            distro_tree = DistroTree.by_id(kwargs['distro_tree_id'][0])
            if not kwargs.get('distro'):
                kwargs['distro'] = distro_tree.distro.name
            if not kwargs.get('osmajor'):
                kwargs['osmajor'] = distro_tree.distro.osversion.osmajor.osmajor

        options = {}
        options['tag'] = [tag.tag for tag in DistroTag.used()]
        options['osmajor'] = [osmajor.osmajor for osmajor in
                OSMajor.ordered_by_osmajor(OSMajor.in_any_lab())]
        options['distro'] = self._get_distro_options(
                osmajor=kwargs.get('osmajor'), tag=kwargs.get('tag'))
        options['distro_tree_id'] = self._get_distro_tree_options(
                distro=kwargs.get('distro'))
        options['lab'] = [lc.fqdn for lc in
                LabController.query.filter(LabController.removed == None)]
        return dict(title=u'Reserve Workflow',
                selection=kwargs, options=options)

    @expose(allow_json=True)
    def get_distro_options(self, **kwargs):
        return {'options': self._get_distro_options(**kwargs)}

    def _get_distro_options(self, osmajor=None, tag=None, system=None, **kwargs):
        """
        Returns a list of distro names for the given osmajor and tag.
        """
        if not osmajor:
            return []
        distros = Distro.query.join(Distro.osversion, OSVersion.osmajor)\
                .filter(Distro.trees.any(DistroTree.lab_controller_assocs.any()))\
                .filter(OSMajor.osmajor == osmajor)\
                .order_by(Distro.date_created.desc())
        if tag:
            distros = distros.filter(Distro._tags.any(DistroTag.tag == tag))
        if system:
            try:
                system = System.by_fqdn(system, identity.current.user)
            except DatabaseLookupError:
                return []
            distros = system.distros(query=distros)
        return [name for name, in distros.values(Distro.name)]

    @expose(allow_json=True)
    def get_distro_tree_options(self, **kwargs):
        return {'options': self._get_distro_tree_options(**kwargs)}

    def _get_distro_tree_options(self, distro=None, system=None, **kwargs):
        """
        Returns a list of distro trees for the given distro.
        """
        if not distro:
            return []
        try:
            distro = Distro.by_name(distro)
        except DatabaseLookupError:
            return []
        trees = distro.dyn_trees.join(DistroTree.arch)\
                .filter(DistroTree.lab_controller_assocs.any())\
                .order_by(DistroTree.variant, Arch.arch)
        if system:
            try:
                system = System.by_fqdn(system, identity.current.user)
            except DatabaseLookupError:
                return []
            trees = system.distro_trees(query=trees)
        return [(tree.id, six.text_type(tree)) for tree in trees]
